src/lib/ares_process.c

```diff
@@ -816,6 +854,77 @@ static struct server_state *ares__random_server(ares_channel_t *channel)
   return NULL;
 }
 
+/* Pick a server from the list with failover behavior.
+ *
+ * We default to using the first server in the sorted list of servers. That is
+ * the server with the lowest number of consecutive failures, and then the
+ * highest priority server (by idx) if there is a draw.
+ *
+ * However, this by itself has two issues:
+ * (1) If a server temporarily goes down and hits some failures, then that
+ *     server will never be retried until all other servers hit the same number
+ *     of failures. This may prevent the server from being retried for a long
+ *     time.
+ * (2) If all servers are failed, then we will continue to select the
+ *     server with the lowest number of consecutive failures even though it may
+ *     be more broken than other servers later in the list.
+ *
+ * For issue (1), with some probability we select a failed server to retry
+ * instead.
+ *
+ * For issue (2), if all servers in the list have a serious failure, then we
+ * fall back to selecting a random server.
+ */
+static struct server_state *ares__failover_server(ares_channel_t *channel)
+{
+  struct server_state *server;
+
+  /* Select the best choice of server, based on consecutive failures and
+   * server idx.
+   */
+  server = ares__slist_first_val(channel->servers);
+
+  /* Defensive code against no servers being available on the channel. */
+  if (server == NULL) {
+    return NULL;
+  }
+
+  if (channel->server_serious_fail_limit > 0 &&
+      server->consec_failures >= channel->server_serious_fail_limit) {
+    /* All servers have a serious failure so fall back to a random server. */
+    return ares__random_server(channel);
+  }
+
```

Comment:
lets get rid of the `server_serious_fail_limit`.  We still want to check servers with failures that haven't hit some random limit, remember, they've been sorted to the end so won't be tried unless other servers have also failed.

Reply:
To give an example of the issue this was trying to solve:

- Suppose we have 2 servers, A and B.
- Initially, both servers are failed. We will alternate query attempts between the two.
  - After 2000 queries A and B will both have 1000 failures.
- Suppose that A recovers.
  - After the next query A will have 0 failures and B will have 1000 failures.
- Suppose that A fails again and B recovers shortly after.
  - It will take 1000 queries sent to A before B is ever retried, which may prolong recovery and cause longer outages.

The serious failure limit was added to give an upper bound on the number of failures before we retry servers randomly.

As an example. setting to `3` would mean if all servers hit `>=3` failures then they are all treated as "equally failed" and chosen from randomly.

Maybe this is too rare of an edge case to care about, especially with the new retry mechanism?

Reply:
I'd think with the new random server failure retry, that alone would resolve the issue without needing a serious failure count concept, or am I misunderstanding something?